In [87]:
import numpy as np
import pandas as pd

In [88]:
train= pd.read_csv(r"E:\ML\train.csv")
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4209 entries, 0 to 4208
Columns: 378 entries, ID to X385
dtypes: float64(1), int64(369), object(8)
memory usage: 12.1+ MB


In [89]:
test =pd.read_csv(r"E:\ML\test.csv")
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4209 entries, 0 to 4208
Columns: 377 entries, ID to X385
dtypes: int64(369), object(8)
memory usage: 12.1+ MB


In [90]:
#train.describe()

In [91]:
#test.describe()

# checking  variance is =0



In [92]:
#train data set
train.var()[train.var() == 0].index.values



array(['X11', 'X93', 'X107', 'X233', 'X235', 'X268', 'X289', 'X290',
       'X293', 'X297', 'X330', 'X347'], dtype=object)

In [93]:
#train data set
train_zero_col=train.var()[train.var() == 0].index.values
train_zero_col

array(['X11', 'X93', 'X107', 'X233', 'X235', 'X268', 'X289', 'X290',
       'X293', 'X297', 'X330', 'X347'], dtype=object)

# Droping the above train columns with variance=0

In [94]:
train_dropped_data=train.drop(train_zero_col, axis=1)

In [95]:
train_dropped_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4209 entries, 0 to 4208
Columns: 366 entries, ID to X385
dtypes: float64(1), int64(357), object(8)
memory usage: 11.8+ MB


In [96]:
# in original train data set the Columns: 378 entries, ID to X385 
# after deleting some columns the Columns: 366 entries, ID to X385

In [97]:
test.var()[test.var() == 0].index.values

array(['X257', 'X258', 'X295', 'X296', 'X369'], dtype=object)

In [98]:
test_zero_col=test.var()[test.var() == 0].index.values
test_zero_col

array(['X257', 'X258', 'X295', 'X296', 'X369'], dtype=object)

In [99]:
test_dropped_data=test.drop(test_zero_col, axis=1)

In [100]:
test_dropped_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4209 entries, 0 to 4208
Columns: 372 entries, ID to X385
dtypes: int64(364), object(8)
memory usage: 11.9+ MB


Columns: 377 entries, ID to X385 
        after dropping
Columns: 372 entries, ID to X385

# Step-2

In [101]:
train_columns=train_dropped_data.columns
train_columns

Index(['ID', 'y', 'X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8',
       ...
       'X375', 'X376', 'X377', 'X378', 'X379', 'X380', 'X382', 'X383', 'X384',
       'X385'],
      dtype='object', length=366)

# checking null values in train Data_set


In [102]:
np.sum(train_dropped_data.isnull())

ID      0
y       0
X0      0
X1      0
X2      0
X3      0
X4      0
X5      0
X6      0
X8      0
X10     0
X12     0
X13     0
X14     0
X15     0
X16     0
X17     0
X18     0
X19     0
X20     0
X21     0
X22     0
X23     0
X24     0
X26     0
X27     0
X28     0
X29     0
X30     0
X31     0
       ..
X355    0
X356    0
X357    0
X358    0
X359    0
X360    0
X361    0
X362    0
X363    0
X364    0
X365    0
X366    0
X367    0
X368    0
X369    0
X370    0
X371    0
X372    0
X373    0
X374    0
X375    0
X376    0
X377    0
X378    0
X379    0
X380    0
X382    0
X383    0
X384    0
X385    0
Length: 366, dtype: int64

# checking null values in test Data_set

In [103]:
np.sum(test_dropped_data.isnull())

ID      0
X0      0
X1      0
X2      0
X3      0
X4      0
X5      0
X6      0
X8      0
X10     0
X11     0
X12     0
X13     0
X14     0
X15     0
X16     0
X17     0
X18     0
X19     0
X20     0
X21     0
X22     0
X23     0
X24     0
X26     0
X27     0
X28     0
X29     0
X30     0
X31     0
       ..
X354    0
X355    0
X356    0
X357    0
X358    0
X359    0
X360    0
X361    0
X362    0
X363    0
X364    0
X365    0
X366    0
X367    0
X368    0
X370    0
X371    0
X372    0
X373    0
X374    0
X375    0
X376    0
X377    0
X378    0
X379    0
X380    0
X382    0
X383    0
X384    0
X385    0
Length: 372, dtype: int64

In [104]:
test_columns=test_dropped_data.columns


# checking unique  values for train data set


In [105]:
train_columns.unique()

Index(['ID', 'y', 'X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8',
       ...
       'X375', 'X376', 'X377', 'X378', 'X379', 'X380', 'X382', 'X383', 'X384',
       'X385'],
      dtype='object', length=366)

In [106]:
#train_columns.apply(pd.value_counts)

# checking uniuqe values for train data set

In [107]:
test_columns.unique()

Index(['ID', 'X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8', 'X10',
       ...
       'X375', 'X376', 'X377', 'X378', 'X379', 'X380', 'X382', 'X383', 'X384',
       'X385'],
      dtype='object', length=372)

In [108]:
#train.apply(pd.value_counts)

In [110]:
train_dropped_data.describe(include=['object'])

X0    X1    X2    X3    X4    X5    X6    X8
count   4209  4209  4209  4209  4209  4209  4209  4209
unique    47    27    44     7     4    29    12    25
top        z    aa    as     c     d     v     g     j
freq     360   833  1659  1942  4205   231  1042   277

In [111]:
test_dropped_data.describe(include=['object'])

X0    X1    X2    X3    X4    X5    X6    X8
count   4209  4209  4209  4209  4209  4209  4209  4209
unique    49    27    45     7     4    32    12    25
top       ak    aa    as     c     d     v     g     e
freq     432   826  1658  1900  4203   246  1073   274

In [112]:
#checking the categorical data in train data set
label_columns = train_dropped_data.describe(include=['object']).columns.values
label_columns



array(['X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8'], dtype=object)

In [113]:

label1_columns = test_dropped_data.describe(include=['object']).columns.values
label1_columns


array(['X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8'], dtype=object)

# # Step:3  Apply label encoder

In [114]:
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder
le = LabelEncoder()
for col in label_columns:
    le.fit(train_dropped_data[col].append(test_dropped_data[col]).values)
    train_dropped_data[col]=le.transform(train_dropped_data[col])
    test_dropped_data[col]=le.transform(test_dropped_data[col])

In [115]:
#features= train_dropped_data.iloc[:,2:].values
#labels = train_dropped_data.iloc[:,-1].values

In [116]:
features.ndim


2

In [195]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(features,
                                                 labels,
                                                 test_size=0.2,
                                               random_state=10)

In [196]:
from sklearn.linear_model import LogisticRegression
leg= LogisticRegression()
leg.fit(X_train,y_train)

C:\Users\thota\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='warn',
          tol=0.0001, verbose=0, warm_start=False)

In [197]:
TrainScore = leg.score(X_train,y_train)
TestScore = leg.score(X_test,y_test)
print("Training Score {} Testing Score {}". format(TrainScore,TestScore))

Training Score 1.0 Testing Score 1.0


# Applying PCA


In [198]:
from sklearn.decomposition import PCA
principalComponents = PCA(n_components=98)
principalComponents.fit(features,labels)

PCA(copy=True, iterated_power='auto', n_components=98, random_state=None,
  svd_solver='auto', tol=0.0, whiten=False)

In [199]:
sklearn_pca = PCA(n_components=0.98)

In [200]:
sklearn_pca.fit(X_train)

PCA(copy=True, iterated_power='auto', n_components=0.98, random_state=None,
  svd_solver='auto', tol=0.0, whiten=False)

In [201]:
principalComponents.explained_variance_ratio_


array([4.08689879e-01, 2.17585077e-01, 1.31200808e-01, 1.07835223e-01,
       8.16524789e-02, 1.40934027e-02, 6.60951171e-03, 3.84659192e-03,
       2.60288938e-03, 2.14378144e-03, 2.09857078e-03, 1.80387564e-03,
       1.48016508e-03, 1.31163426e-03, 1.14048800e-03, 9.18564405e-04,
       8.49319384e-04, 7.95592011e-04, 6.96370320e-04, 6.28377000e-04,
       5.72268089e-04, 5.53034525e-04, 5.18552066e-04, 4.54851774e-04,
       4.30273667e-04, 3.83978619e-04, 3.66708831e-04, 3.34660628e-04,
       3.27499427e-04, 3.15393259e-04, 2.93055335e-04, 2.78599266e-04,
       2.65173410e-04, 2.51329158e-04, 2.37484612e-04, 2.20153218e-04,
       1.96197354e-04, 1.90392465e-04, 1.83388549e-04, 1.73952729e-04,
       1.65849825e-04, 1.58461956e-04, 1.55772773e-04, 1.54537011e-04,
       1.51383347e-04, 1.35778337e-04, 1.32427569e-04, 1.27356343e-04,
       1.25380242e-04, 1.22367374e-04, 1.17783437e-04, 1.14937200e-04,
       1.11561713e-04, 1.05705421e-04, 1.01665644e-04, 9.57661378e-05,
      

In [202]:

pca_x_train = pd.DataFrame(principalComponents.transform(X_train))
pca_x_train.head()

0          1          2          3          4         5         6   \
0 -17.460399   1.935352   2.905539  -7.325726   6.305020  4.145486  1.727192   
1 -11.167472  -8.250288   4.584157  13.363837  -4.433659  3.105982  1.932175   
2  10.424728 -11.392937 -10.210733   2.332987  -7.703544 -1.388951 -1.989637   
3  23.300929  -7.147855 -10.330900   0.474658   6.089487 -1.198755 -2.628153   
4 -20.425201  -6.652617   9.775983 -10.331518 -11.222868 -4.283098  1.999772   

         7         8         9   ...        88        89        90        91  \
0 -2.658276  0.859078  0.801523  ...  0.045232 -0.045912 -0.051192  0.042600   
1  1.901826  0.062763  1.152018  ...  0.040010 -0.006988 -0.014692 -0.086458   
2  0.495455 -0.915177 -0.234765  ... -0.135635 -0.189466  0.241246  0.023604   
3 -1.689854 -0.342056 -0.296889  ... -0.083641 -0.019309 -0.112937 -0.071756   
4  1.474038 -0.813903  0.600996  ... -0.091381 -0.107056  0.147325 -0.173430   

         92        93        94        95        96        97  
0 -0.067003 -0.040685  0.040446 -0.037943 -0.000974 -0.035866  
1  0.023747 -0.005745  0.007715  0.028584  0.045622  0.023303  
2  0.246032  0.002719 -0.125410 -0.325507 -0.108915  0.491750  
3  0.040901  0.013947 -0.077491  0.014582 -0.017744  0.051971  
4  0.057080 -0.003879 -0.020524  0.024307  0.056571  0.025615  

[5 rows x 98 columns]

In [203]:
pca_test=pd.DataFrame(principalComponents.transform(X_test))
pca_test.head()

0          1         2          3          4         5         6   \
0 -12.199529   1.473405  4.969911 -10.136803 -10.512450  4.961853 -0.606839   
1 -15.090736  -8.738195  5.151048  -9.053640   2.948246 -4.046263 -0.282178   
2  10.155512  -5.595474 -8.128344   4.727639   0.327876  6.633871 -1.304357   
3  30.925451  15.620637 -6.479820   7.688562  -2.190709  3.768786  3.551019   
4  24.917387  -7.209175  5.749065   2.208805  -3.585600  2.424282 -2.147924   

         7         8         9   ...        88        89        90        91  \
0 -2.489584  0.679529  0.289930  ...  0.055950  0.033347 -0.114377  0.042429   
1  0.870825 -0.497458 -0.285826  ...  0.082133 -0.118290  0.145265  0.124377   
2  1.873477  0.406918 -0.265205  ...  0.392471  0.333267 -0.545884 -0.056591   
3 -0.213789  0.438899 -1.114165  ... -0.000520  0.049309 -0.052713  0.002438   
4 -1.611703 -0.950562 -0.123950  ... -0.036924 -0.087115  0.027092  0.039931   

         92        93        94        95        96        97  
0  0.021405 -0.053791 -0.108753 -0.013734  0.053120  0.033189  
1 -0.081129  0.014491 -0.049435  0.044611 -0.070764 -0.018991  
2  0.494265  0.416335  0.042495 -0.656134  0.066815 -0.577203  
3 -0.036450  0.033468  0.006046  0.070276 -0.034649  0.018212  
4  0.094153  0.084832 -0.067640 -0.001276 -0.073770 -0.061408  

[5 rows x 98 columns]

In [204]:
print(finalFeatures.shape)

(3367, 98)


In [205]:
from sklearn.neighbors import KNeighborsClassifier
model = KNeighborsClassifier()
model.fit(X_train,y_train)


KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=None, n_neighbors=5, p=2,
           weights='uniform')

In [206]:
print(model.score(X_train,y_train))
print(model.score(X_test,y_test))

0.9994059994059994
1.0


In [207]:
principalComponents.components_ 


array([[-9.47882068e-01,  2.09264757e-01,  2.29661509e-01, ...,
         7.06159720e-06, -1.00973150e-05, -2.62563493e-05],
       [ 2.35547968e-01, -4.50142622e-03,  9.68073720e-01, ...,
        -8.85958568e-05,  2.74561489e-08, -1.08927086e-05],
       [ 5.65297939e-02,  2.80718825e-01,  1.52970181e-02, ...,
        -7.31904544e-05,  3.07685011e-05,  1.50975422e-04],
       ...,
       [ 4.22664080e-04,  2.37064850e-03,  5.24263708e-03, ...,
        -1.07578219e-02,  3.79412087e-04,  5.31751941e-02],
       [ 3.79075753e-04,  1.64934973e-03, -5.32473343e-03, ...,
         2.57516934e-02, -4.03399571e-04,  2.49053341e-02],
       [-1.08947425e-03,  1.77615382e-03,  1.46877895e-03, ...,
        -2.33862283e-03, -2.38301034e-03,  9.08159913e-03]])

# xg boost

In [208]:
from  xgboost import XGBRFRegressor

In [209]:
model= XGBRFRegressor(object='reg:linear',learning_rate=0.1)

In [210]:
model=XGBRFRegressor()
model.fit(pca_x_train,y_train)

[13:38:56] WARNING: C:/Jenkins/workspace/xgboost-win64_release_0.90/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.


XGBRFRegressor(base_score=0.5, colsample_bylevel=1, colsample_bynode=0.8,
        colsample_bytree=1, gamma=0, learning_rate=1, max_delta_step=0,
        max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
        n_jobs=1, nthread=None, objective='reg:linear', random_state=0,
        reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
        silent=None, subsample=0.8, verbosity=1)

In [211]:
print(model.score(pca_x_train,y_train))


0.9747845225621989


1) The  projrct has ended upto here 